File: src/ipc/commands.ts
import { invoke, isBrowserMode } from "./invoke";
import { CcoDocs, CcoShorthand, CharacterDb, ContextDb, ImageStatus, RawElement, RoundtripReport, TwuiDocument } from "../types/twui";

export function loadContextDb(): Promise<ContextDb> {
  return invoke("load_context_db");
}

/** Character generation templates + resolved portrait folders (Characters panel). */
export function loadCharacterDb(): Promise<CharacterDb> {
  return invoke("load_character_db");
}

/** CCO symbol table parsed from the game's UI documentation (Inspector hints). */
export function loadCcoDocs(): Promise<CcoDocs> {
  return invoke("load_cco_docs");
}

/** Content-defined CCO shorthand macros (`ui/cco/*.json`), keyed by CCO type. */
export function loadCcoShorthand(): Promise<CcoShorthand> {
  return invoke("load_cco_shorthand");
}

export function loadTemplates(ids: string[]): Promise<Record<string, TwuiDocument>> {
  return invoke("load_templates", { ids });
}

/** Load layouts referenced by ComponentCreator (arbitrary data-root-relative paths). */
export function loadLayouts(paths: string[]): Promise<Record<string, TwuiDocument>> {
  return invoke("load_layouts", { paths });
}

/** Localised UI strings keyed by bare record key (campaign_localised_strings). */
export function loadLoc(): Promise<Record<string, string>> {
  return invoke("load_loc");
}

export interface ScriptHit {
  /** Path relative to the data root, forward-slashed. */
  path: string;
  text: string;
}

/** Locate the Lua script that backs a panel's script_id (set_context_value). */
export function findScript(scriptId: string): Promise<ScriptHit | null> {
  return invoke("find_script", { scriptId });
}

/** Read a .lua script file (sandboxed to the data root). */
export function readScript(path: string): Promise<string> {
  return invoke("read_script", { path });
}

export function serializeElement(element: RawElement): Promise<string> {
  return invoke("serialize_element", { element });
}

export function parseElement(text: string): Promise<RawElement> {
  return invoke("parse_element", { text });
}

export function listBackgrounds(): Promise<string[]> {
  return invoke("list_backgrounds");
}

export function getDataRoot(): Promise<string | null> {
  return invoke("get_data_root");
}

export function setDataRoot(path: string): Promise<void> {
  return invoke("set_data_root", { path });
}

/** Names of the games under the `games/` directory (3K, WH3, …). */
export function listGames(): Promise<string[]> {
  return invoke("list_games");
}

/** The active game name (data root folder under `games/`), or null for a custom root. */
export function currentGame(): Promise<string | null> {
  return invoke("current_game");
}

/** Switch the active game by name (a subfolder of `games/`). */
export function setGame(name: string): Promise<void> {
  return invoke("set_game", { name });
}

/** Switch to pack mode: read `.pack` files under `gameDir` (read-only).
 *  `includeMods=false` loads only vanilla (non-Mod-type) packs. */
export function setPackSource(gameDir: string, includeMods: boolean): Promise<void> {
  return invoke("set_pack_source", { gameDir, includeMods });
}

/** True when the active source is `.pack` archives (vs a loose folder). */
export function isPackMode(): Promise<boolean> {
  return invoke("is_pack_mode");
}

/** Every `.twui.xml` reachable from the active source (pack content browser). */
export function listLayouts(): Promise<string[]> {
  return invoke("list_layouts");
}

/** Every image (png/dds/tga/jpg) in the active source — backs the image finder. */
export function listImages(): Promise<string[]> {
  return invoke("list_images");
}

/** Overlay a single `.pack` (absolute path) over the active source. */
export function setOverlayPack(path: string): Promise<void> {
  return invoke("set_overlay_pack", { path });
}

/** Remove the single-pack overlay, restoring the base source. */
export function clearOverlayPack(): Promise<void> {
  return invoke("clear_overlay_pack");
}

/** The configured RPFM `.ron` schema path (decodes binary db tables), or null. */
export function getSchemaPath(): Promise<string | null> {
  return invoke("get_schema_path");
}

/** Point at the user's local RPFM `.ron` schema file (e.g. schema_3k.ron). */
export function setSchemaPath(path: string): Promise<void> {
  return invoke("set_schema_path", { path });
}

/** Read+parse a layout by source-relative path (open from the pack browser). */
export function readLayoutRel(rel: string): Promise<TwuiDocument> {
  return invoke("read_layout_rel", { rel });
}

export function readLayout(path: string): Promise<TwuiDocument> {
  return invoke("read_layout", { path });
}

export function saveLayout(path: string, doc: TwuiDocument): Promise<void> {
  return invoke("save_layout", { path, doc });
}

export function roundtripCheck(path: string): Promise<RoundtripReport> {
  return invoke("roundtrip_check", { path });
}

export function imageStatus(imagePath: string): Promise<ImageStatus> {
  return invoke("image_status", { imagePath });
}

/** An image passed inline to the bug-report webhook (program shot / visualizer render).
 *  `b64` may be a bare base64 string or a full `data:image/png;base64,...` data URL. */
export interface InlineImage {
  name: string;
  b64: string;
}

/** A bug report submitted from the in-app menu; mirrors the Rust `BugReport` struct. */
export interface BugReport {
  description: string;
  contact?: string;
  /** Diagnostic key/values (app version, OS, game, file, resolution) shown as embed fields. */
  meta: Record<string, string | number | null>;
  inlineImages: InlineImage[];
  /** Absolute paths to user-picked images, read by the backend at submit time. */
  filePaths: string[];
}

/** Capture the whole app window; returns a `data:image/png;base64,...` URL. */
export function captureAppWindow(): Promise<string> {
  return invoke("capture_app_window");
}

/** Deliver a bug report (description + images) to the author's configured Discord webhook. */
export function submitBugReport(report: BugReport): Promise<void> {
  return invoke("submit_bug_report", { report });
}

// On Windows (and Android) Tauri serves custom URI schemes as
// `http://<scheme>.localhost/...`; on macOS/Linux as `<scheme>://localhost/...`.
const IS_WINDOWS = typeof navigator !== "undefined" && /Windows/i.test(navigator.userAgent);

/** Build a URL for a TWUI imagepath. In the Tauri webview this uses the twuiimg:// protocol;
 *  in the web client it uses the same-origin `/img/{rel}` HTTP route. `bust` (an image epoch) is
 *  appended as a `?v=` query so a source change defeats the HTTP cache; the backend resolves
 *  images from the URI path only and ignores the query. */
export function imageUrl(relPath: string, bust?: string | number): string {
  // encodeURI keeps '/' but encodes spaces etc.; backend percent-decodes.
  const enc = encodeURI(relPath);
  const q = bust !== undefined ? `?v=${encodeURIComponent(String(bust))}` : "";
  if (isBrowserMode()) {
    return `/img/${enc}${q}`;
  }
  return IS_WINDOWS
    ? `http://twuiimg.localhost/${enc}${q}`
    : `twuiimg://localhost/${enc}${q}`;
}

// --- Host file browser (web client uses this in place of native OS dialogs) ---

/** One entry in a host directory listing. Field names mirror the Rust `DirEntry`. */
export interface HostDirEntry {
  name: string;
  path: string;
  is_dir: boolean;
}

/** A host directory listing (or the drive/root list when `path` is null). */
export interface HostDirListing {
  path: string | null;
  parent: string | null;
  entries: HostDirEntry[];
}

/** Suggested starting directories for the host file browser. */
export interface HostPaths {
  data_root: string | null;
  games_dir: string | null;
}

/** List a host directory; pass null/undefined to list filesystem roots (drives). */
export function hostListDir(path?: string | null): Promise<HostDirListing> {
  return invoke("host_list_dir", { path: path ?? null });
}

export function hostDefaultPaths(): Promise<HostPaths> {
  return invoke("host_default_paths");
}

// --- Web access point control (desktop only; experimental) ---

export type WebBind = "loopback" | "lan" | "custom";

export interface WebOpts {
  bind: WebBind;
  port: number;
  customIp?: string | null;
  password: string;
}

export interface WebInfo {
  url: string;
  bind: WebBind;
  host: string;
  port: number;
}

/** Start the web access server; returns a share URL. Errors (e.g. port in use) reject. */
export function startWebServer(opts: WebOpts): Promise<WebInfo> {
  return invoke("start_web_server", { opts });
}

export function stopWebServer(): Promise<void> {
  return invoke("stop_web_server");
}

/** Current web server info, or null when stopped. */
export function webServerStatus(): Promise<WebInfo | null> {
  return invoke("web_server_status");
}

File: src-tauri/src/commands.rs
//! Tauri command surface exposed to the desktop webview. Each command is a thin
//! shim over `crate::api`, which holds the actual logic and is shared with the
//! experimental web server (see `web/`). Keep behavior changes in `api.rs`.

use crate::api::{self, RoundtripReport};
use crate::character::CharacterDb;
use crate::cco_docs::CcoDocs;
use crate::cco_shorthand::CcoShorthand;
use crate::db::ContextDb;
use crate::image::ImageStatus;
use crate::model::{Document, Element};
use crate::script::ScriptHit;
use crate::state::AppState;
use tauri::State;

/// Serialize a single component/hierarchy element to UIEd-style XML text
/// (reuses the document serializer at depth 0; output matches Save).
#[tauri::command]
pub fn serialize_element(element: Element) -> String {
    api::serialize_element(element)
}

/// Parse a single `<tag …>…</tag>` fragment back into an element.
#[tauri::command]
pub fn parse_element(text: String) -> Result<Element, String> {
    api::parse_element(&text)
}

/// List selectable background images under `background/`.
#[tauri::command]
pub fn list_backgrounds(state: State<AppState>) -> Vec<String> {
    api::list_backgrounds(&state)
}

#[tauri::command]
pub fn load_context_db(state: State<AppState>) -> ContextDb {
    api::load_context_db(&state)
}

/// Load the character generation templates and their resolved portrait folders,
/// for the Characters panel (assigning characters to a screen's roles).
#[tauri::command]
pub fn load_character_db(state: State<AppState>) -> CharacterDb {
    api::load_character_db(&state)
}

/// Parse the CCO symbol table from the game's UI documentation (Inspector hints).
#[tauri::command]
pub fn load_cco_docs(state: State<AppState>) -> CcoDocs {
    api::load_cco_docs(&state)
}

/// Load the content-defined CCO shorthand macros (`ui/cco/*.json`) — named
/// expressions components reference by name in `context_function_id`.
#[tauri::command]
pub fn load_cco_shorthand(state: State<AppState>) -> CcoShorthand {
    api::load_cco_shorthand(&state)
}

/// Load localised UI strings (campaign_localised_strings) keyed by bare record
/// key — used to show real text where the `.twui.xml` only has a loc label.
#[tauri::command]
pub fn load_loc(state: State<AppState>) -> std::collections::HashMap<String, String> {
    api::load_loc(&state)
}

/// Locate the Lua script backing a panel's `script_id` (the file that calls
/// `set_context_value("<id>", …)`). Returns its data-root-relative path + text.
#[tauri::command]
pub fn find_script(state: State<AppState>, script_id: String) -> Option<ScriptHit> {
    api::find_script(&state, &script_id)
}

/// Read a `.lua` script file (sandboxed to the data root) for manual connection.
#[tauri::command]
pub fn read_script(state: State<AppState>, path: String) -> Result<String, String> {
    api::read_script(&state, &path)
}

/// Read a DB table (header + stringified rows) for the preview-binding feature.
/// Missing/undecodable table -> empty header+rows.
#[tauri::command]
pub fn read_db_table(state: State<AppState>, table: String) -> api::DbTable {
    api::read_db_table(&state, &table)
}

/// Read and parse template layouts referenced by `template_id`, keyed by id.
#[tauri::command]
pub fn load_templates(
    state: State<AppState>,
    ids: Vec<String>,
) -> std::collections::HashMap<String, Document> {
    api::load_templates(&state, ids)
}

/// Read and parse layouts referenced by `ComponentCreator` (arbitrary paths under
/// the data root), keyed by the given path.
#[tauri::command]
pub fn load_layouts(
    state: State<AppState>,
    paths: Vec<String>,
) -> std::collections::HashMap<String, Document> {
    api::load_layouts(&state, paths)
}

#[tauri::command]
pub fn get_data_root(state: State<AppState>) -> Option<String> {
    api::get_data_root(&state)
}

/// Names of the games available under the `games/` directory (3K, WH3, …).
#[tauri::command]
pub fn list_games() -> Vec<String> {
    api::list_games()
}

/// The currently-selected game name (the data root's folder, when it's a game
/// under `games/`); None for a custom/legacy data root.
#[tauri::command]
pub fn current_game(state: State<AppState>) -> Option<String> {
    api::current_game(&state)
}

/// Switch the active game by name (a subfolder of `games/`).
#[tauri::command]
pub fn set_game(state: State<AppState>, name: String) -> Result<(), String> {
    api::set_game(&state, &name)
}

/// Update the active rpfm game key + drop game-specific caches, without re-pointing the data root.
#[tauri::command]
pub fn set_game_key(state: State<AppState>, name: String) -> Result<(), String> {
    api::set_game_key(&state, &name)
}

#[tauri::command]
pub fn set_data_root(state: State<AppState>, path: String) -> Result<(), String> {
    api::set_data_root(&state, &path)
}

/// Switch to pack mode: read the `.pack` files under `game_dir` (read-only).
/// `include_mods=false` loads only vanilla (non-Mod-type) packs.
#[tauri::command]
pub fn set_pack_source(
    state: State<AppState>,
    game_dir: String,
    include_mods: bool,
) -> Result<(), String> {
    api::set_pack_source(&state, &game_dir, include_mods)
}

/// True when the active source is `.pack` archives (vs a loose folder).
#[tauri::command]
pub fn is_pack_mode(state: State<AppState>) -> bool {
    api::is_pack_mode(&state)
}

/// Every `.twui.xml` reachable from the active source, as relative paths.
#[tauri::command]
pub fn list_layouts(state: State<AppState>) -> Vec<String> {
    api::list_layouts(&state)
}

/// Every image (png/dds/tga/jpg) reachable from the active source.
#[tauri::command]
pub fn list_images(state: State<AppState>) -> Vec<String> {
    api::list_images(&state)
}

/// Overlay a single `.pack` over the active source (reads resolve from it first,
/// then fall back). The path is an absolute file path from the OS dialog.
#[tauri::command]
pub fn set_overlay_pack(state: State<AppState>, path: String) -> Result<(), String> {
    api::set_overlay_pack(&state, &path)
}

/// Remove the single-pack overlay, restoring the base source.
#[tauri::command]
pub fn clear_overlay_pack(state: State<AppState>) {
    api::clear_overlay_pack(&state)
}

/// The active single-pack overlay path, or None.
#[tauri::command]
pub fn get_overlay_pack(state: State<AppState>) -> Option<String> {
    api::get_overlay_pack(&state)
}

/// The configured RPFM `.ron` schema path (decodes binary db tables), or None.
#[tauri::command]
pub fn get_schema_path(state: State<AppState>) -> Option<String> {
    api::get_schema_path(&state)
}

/// Point at the user's local RPFM `.ron` schema file (e.g. `schema_3k.ron`).
#[tauri::command]
pub fn set_schema_path(state: State<AppState>, path: String) -> Result<(), String> {
    api::set_schema_path(&state, &path)
}

/// Read+parse a layout by source-relative path (used to open files from the
/// pack content browser; folder mode resolves under the data root).
#[tauri::command]
pub fn read_layout_rel(state: State<AppState>, rel: String) -> Result<Document, String> {
    api::read_layout_rel(&state, &rel)
}

#[tauri::command]
pub fn read_layout(path: String) -> Result<Document, String> {
    api::read_layout(&path)
}

#[tauri::command]
pub fn save_layout(path: String, doc: Document) -> Result<(), String> {
    api::save_layout(&path, doc)
}

/// Write a PNG (base64 or `data:image/png;base64,...`) to an absolute host path.
#[tauri::command]
pub fn save_png(path: String, b64: String) -> Result<(), String> {
    api::save_png(&path, &b64)
}

// --- Editable pack workspace (the "Pack Editor") ---

/// Create a new empty Mod pack at `path` and open it as the editable workspace.
#[tauri::command]
pub fn new_pack_workspace(state: State<AppState>, path: String) -> Result<(), String> {
    api::new_pack_workspace(&state, &path)
}

/// Open an existing `.pack` as the editable workspace; returns its `.twui.xml` paths.
#[tauri::command]
pub fn open_pack_workspace(state: State<AppState>, path: String) -> Result<Vec<String>, String> {
    api::open_pack_workspace(&state, &path)
}

/// The `.twui.xml` paths in the open workspace pack.
#[tauri::command]
pub fn list_workspace_layouts(state: State<AppState>) -> Vec<String> {
    api::list_workspace_layouts(&state)
}

/// Read+parse a layout from the workspace pack.
#[tauri::command]
pub fn read_workspace_layout(state: State<AppState>, rel: String) -> Result<Document, String> {
    api::read_workspace_layout(&state, &rel)
}

/// Write `doc` into the workspace pack (create/replace `rel`) and persist the pack to disk.
#[tauri::command]
pub fn save_workspace_layout(
    state: State<AppState>,
    rel: String,
    doc: Document,
) -> Result<(), String> {
    api::save_workspace_layout(&state, &rel, doc)
}

/// Delete a layout from the workspace pack and persist.
#[tauri::command]
pub fn delete_workspace_layout(state: State<AppState>, rel: String) -> Result<(), String> {
    api::delete_workspace_layout(&state, &rel)
}

/// Close the editable workspace.
#[tauri::command]
pub fn close_pack_workspace(state: State<AppState>) {
    api::close_pack_workspace(&state)
}

/// The open workspace's path + dirty flag.
#[tauri::command]
pub fn pack_workspace_status(state: State<AppState>) -> api::WorkspaceStatus {
    api::pack_workspace_status(&state)
}

/// Round-trip a file in memory and report whether serialize(parse(x)) == x.
/// Useful as a fidelity self-check from the UI.
#[tauri::command]
pub fn roundtrip_check(path: String) -> Result<RoundtripReport, String> {
    api::roundtrip_check(&path)
}

#[tauri::command]
pub fn image_status(state: State<AppState>, image_path: String) -> ImageStatus {
    api::image_status(&state, &image_path)
}

// --- Host file browser (used by the web client in place of native dialogs) ---

/// List a host directory (or filesystem roots when `path` is omitted). Only
/// surfaced over the web access point, but registered for both transports.
#[tauri::command]
pub fn host_list_dir(path: Option<String>) -> Result<api::DirListing, String> {
    api::host_list_dir(path.as_deref())
}

/// Suggested starting directories for the host file browser.
#[tauri::command]
pub fn host_default_paths(state: State<AppState>) -> api::HostPaths {
    api::host_default_paths(&state)
}
